README
NAME
    Object::Serializer - General Purpose Object Serializer

VERSION
    version 0.000010

SYNOPSIS
        package Point;

        use Moo;
        use parent 'Object::Serializer';

        has 'x' => (is => 'rw');
        has 'y' => (is => 'rw');

        package main;

        my $point = Point->new(x => 10, y => 10);

        # serialize the class instance into a hash
        my $hash = $point->serialize; # { __CLASS__ => 'Point', x => 10, y => 10 }

DESCRIPTION
    Getting objects into an ideal format for passing representations in and
    out of applications can be a real pain. Object::Serializer is a fast and
    simple pure-perl framework-agnostic type-less none-opinionated
    light-weight primitive general purpose object serializer which tries to
    help make object serialization easier. This module is useful in
    situations when you have blessed objects you wish to produce hash
    representations from which you can store directly or convert to JSON,
    YAML, or XML. This module does not currently support deserialization.

METHODS
  serialize
    The serialize method expects an object and returns a serialized
    (hashified) version of that object.

        my $hash = $self->serialize;
        my $hash = $self->serialize($object);
        my $hash = $self->serialize($object, marker => undef); # no marker

  serialization_strategy_for
    The serialization_strategy_for method expects a reftype and a
    sub-routine. This method registers a custom serialization strategy which
    will be used during the collapsing of the reference type specified.

        CLASS->serialization_strategy_for(
            REFTYPE => sub { ... }
        );

EXTENSION
    Object::Serializer can be used as a serializer independently, however,
    it is primarily designed to be used as a base class for your classes or
    roles. By default, Object::Serializer doesn't do anything special for
    you in the way of serialization, in-fact, it is little more than a
    wrapper around Data::Dumper. Additionally, you can hook into the
    serialization process by defining your serialization strategy using your
    own custom serialization routines which will be executed whenever a
    specific reference type is encountered. The following syntax is what you
    might use to register your own custom serialization strategy. This
    example registers a custom serializer that is executed globally whenever
    a DateTime object is found.

        Object::Serializer->serialization_strategy_for(
            DateTime => sub { pop->iso8601 }
        );

    Additionally, you can register a serialization strategy to be used only
    when invoked by a specific class. The following syntax is what you might
    use to register a serialization strategy to be executed only for a
    specific class:

        Point->serialization_strategy_for(
            DateTime => sub { pop->iso8601 }
        );

CAVEATS
    Circular references are problematic and should be avoided, you can
    weaken or otherwise handle them yourself then re-assemble them later as
    a means toward getting around this.

AUTHOR
    Al Newkirk <anewkirk@ana.io>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Al Newkirk.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2012-11-11 00:00:00 +0000 to present.
==================================================

---------------------------------------------
version 0.000010 at 2013-11-11 08:41:14 +0000
---------------------------------------------

  Change: e8b9ddd2cd84417f8906cc0138f331a0d1d134a6
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-11-11 03:41:06 +0000

    added strict and warnings and ignores them where appropriate 

---------------------------------------------
version 0.000009 at 2013-11-11 08:24:58 +0000
---------------------------------------------

  Change: 05cb1f3a6a9d6a4b0c3aaaba0628f193b655a1f4
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-11-11 03:24:58 +0000

    v0.000009 

  Change: 92a901e7c8df14a09f3ba9cb3c975a5a2cb9ca67
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-11-11 03:24:39 +0000

    refactored, less code, better execution 

---------------------------------------------
version 0.000008 at 2013-11-11 05:31:56 +0000
---------------------------------------------

  Change: 11036cd8b28f0232b38e6da332de261d614e264c
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-11-11 00:31:56 +0000

    v0.000008 

  Change: 248b03d7a421ca982a3b695a12bb1dfcaf43454b
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-11-11 00:31:17 +0000

    removed deserialization functionality 

---------------------------------------------
version 0.000007 at 2013-09-16 01:41:47 +0000
---------------------------------------------

  Change: d55f16a94091706b93de767ebc061837cc3f54f2
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-15 21:41:47 +0000

    v0.000007 

  Change: 47034ae007aaec3a94c1c4d154caed62b289fdb9
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-15 21:40:47 +0000

    updated Changes 

  Change: c3489995d7c6147f70ee04c7ac8b729f6417cb06
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-15 21:39:53 +0000

    ability to tag during serialization/deserialization process 

---------------------------------------------
version 0.000006 at 2013-09-15 20:17:22 +0000
---------------------------------------------

  Change: 4d08ef4b675e9795c0a7d4b2c192a9683e3e7294
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-15 16:17:22 +0000

    v0.000006 

  Change: ca1857466994c59880164a694245fbc1f0647bd6
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-15 16:17:07 +0000

    added the ability to serialize without tagging 

  Change: 4dbac6b3c1f6219070fb8afc2e696eb91fe490a6
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-15 05:04:15 +0000

    miscellaneous changes (not worth mentioning) 

---------------------------------------------
version 0.000005 at 2013-09-14 22:08:40 +0000
---------------------------------------------

  Change: 4989d761d07070f5fd85054bf6a0a49c6f708c1f
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 18:08:40 +0000

    v0.000005 

  Change: 892e35f6db91236b253a803606c5574e8b4d6c9b
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 18:08:29 +0000

    updated documentation 

---------------------------------------------
version 0.000004 at 2013-09-14 21:46:21 +0000
---------------------------------------------

  Change: a8ffdb314947a8e224884a86a70c49d275f372e7
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 17:46:21 +0000

    v0.000004 

  Change: ab6edbc384ac036e28dba4e319e54b2c572efefc
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 17:46:08 +0000

    new approach toward hashifying objects, removed Regexp::Grammars
    dependency 

---------------------------------------------
version 0.000003 at 2013-09-14 06:30:16 +0000
---------------------------------------------

  Change: 2949f6e06101ac1e98d9478dd1b28e84ed48f4ea
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:30:16 +0000

    v0.000003 

  Change: 94448b9de0686350d86634b0f741ea020c7cdd85
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:29:54 +0000

    update Changes 

  Change: aa66aa6ee8e708d6746171b6d550a2e126d0dd11
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:28:07 +0000

    initial release 

---------------------------------------------
version 0.000002 at 2013-09-14 06:09:09 +0000
---------------------------------------------

  Change: ef809bc489fc0508ae95eb38bb51cd4e4de71376
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:09:09 +0000

    v0.000002 

  Change: f47c01ef8696d51bd71115f9e8fa3866a0ef7e59
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:08:17 +0000

    fixed more typos and dependency misspelling 

---------------------------------------------
version 0.000001 at 2013-09-14 06:01:53 +0000
---------------------------------------------

  Change: 740256ef73a61e18a28fecf2536917894b275fea
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:01:53 +0000

    v0.000001 

  Change: eb67c8d92ece729466ff154c86f5294264c6aaed
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:01:41 +0000

    created Changes 

  Change: fab6b09ccb4badec2a32d9e81de5d1cbbe776729
  Author: Al Newkirk <anewkirk@ana.io>
  Date : 2013-09-14 02:00:36 +0000

    initial commit 

================
End of releases.
================


lib/Object/Serializer.pm
# ABSTRACT: General Purpose Object Serializer
package Object::Serializer;

use utf8;
use 5.010;
use strict;
use warnings;
use Data::Dumper ();
use Scalar::Util qw(blessed refaddr);
our %TYPES;

our $VERSION = '0.000010'; # VERSION


sub new {
    bless {}, shift
}

sub _serialization {
    my ($self, $options, $reference, $class) = @_;

    my $coercion;
    for my $registry (ref $self, __PACKAGE__) {
        my $type = $TYPES{$registry};
        next unless 'HASH' eq ref $type;
        $coercion = $type->{$class};
        return $coercion->(bless $reference, $class)
            if 'CODE' eq ref $coercion;
    }

    # tag blessed hash references
    $reference->{$options->{marker}} = $class
        if defined $options->{marker} && 'HASH' eq ref $reference;

    return $reference;
}


sub serialize {
    my ($self, $object, %options) = @_;

    local $Data::Dumper::Terse      = 1;
    local $Data::Dumper::Indent     = 0;
    local $Data::Dumper::Useqq      = 1;
    local $Data::Dumper::Deparse    = 1;
    local $Data::Dumper::Quotekeys  = 0;
    local $Data::Dumper::Sortkeys   = 1;
    local $Data::Dumper::Deepcopy   = 1;
    local $Data::Dumper::Purity     = 0;

    my $options   = {marker => '__CLASS__', %options};
    my $execution = join '::', __PACKAGE__, 'Runtime', refaddr $self;

    no strict 'refs';
    no warnings 'redefine';

    *{$execution} = sub {
        my @arguments = @_;
        Object::Serializer::_serialization($self, $options, @arguments);
    };

    (my $target = Data::Dumper::Dumper($object // $self)) =~
        s/bless(?=(?:(?:(?:[^"\\]++|\\.)*+"){2})*+(?:[^"\\]++|\\.)*+$)/$execution/g;
    my $hash = do { no strict; eval "my \$VAR1 = $target\n" } or die $@;

    return $hash;
}


sub serialization_strategy_for {
    my ($self, $reftype, $routine) = @_;

    die "Couldn't register reftype serialization strategy ".
        "due to invalid arguments" unless
        $self && $reftype && 'CODE' eq ref $routine
    ;

    return $TYPES{ref($self) // $self}{$reftype} = $routine;
}


1;

__END__

=pod

=head1 NAME

Object::Serializer - General Purpose Object Serializer

=head1 VERSION

version 0.000010

=head1 SYNOPSIS

    package Point;

    use Moo;
    use parent 'Object::Serializer';

    has 'x' => (is => 'rw');
    has 'y' => (is => 'rw');

    package main;

    my $point = Point->new(x => 10, y => 10);

    # serialize the class instance into a hash
    my $hash = $point->serialize; # { __CLASS__ => 'Point', x => 10, y => 10 }

=head1 DESCRIPTION

Getting objects into an ideal format for passing representations in and
out of applications can be a real pain. Object::Serializer is a fast and simple
pure-perl framework-agnostic type-less none-opinionated light-weight primitive
general purpose object serializer which tries to help make object serialization
easier. This module is useful in situations when you have blessed objects you
wish to produce hash representations from which you can store directly or
convert to JSON, YAML, or XML. This module does not currently support
deserialization.

=head1 METHODS

=head2 serialize

The serialize method expects an object and returns a serialized (hashified)
version of that object.

    my $hash = $self->serialize;
    my $hash = $self->serialize($object);
    my $hash = $self->serialize($object, marker => undef); # no marker

=head2 serialization_strategy_for

The serialization_strategy_for method expects a reftype and a sub-routine. This
method registers a custom serialization strategy which will be used during the
collapsing of the reference type specified.

    CLASS->serialization_strategy_for(
        REFTYPE => sub { ... }
    );

=head1 EXTENSION

Object::Serializer can be used as a serializer independently, however, it is
primarily designed to be used as a base class for your classes or roles. By
default, Object::Serializer doesn't do anything special for you in the way of
serialization, in-fact, it is little more than a wrapper around L<Data::Dumper>.
Additionally, you can hook into the serialization process by defining your
serialization strategy using your own custom serialization routines which will
be executed whenever a specific reference type is encountered. The following
syntax is what you might use to register your own custom serialization strategy.
This example registers a custom serializer that is executed globally whenever a
DateTime object is found.

    Object::Serializer->serialization_strategy_for(
        DateTime => sub { pop->iso8601 }
    );

Additionally, you can register a serialization strategy to be used only when
invoked by a specific class. The following syntax is what you might use to
register a serialization strategy to be executed only for a specific class:

    Point->serialization_strategy_for(
        DateTime => sub { pop->iso8601 }
    );

=head1 CAVEATS

Circular references are problematic and should be avoided, you can weaken or
otherwise handle them yourself then re-assemble them later as a means toward
getting around this.

=head1 AUTHOR

Al Newkirk <anewkirk@ana.io>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Al Newkirk.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "General Purpose Object Serializer",
  "AUTHOR" => "Al Newkirk <anewkirk\@ana.io>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Object-Serializer",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Object::Serializer",
  "PREREQ_PM" => {
    "Data::Dumper" => 0,
    "Scalar::Util" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.000010",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



